src/zope/interface/adapter.py

```diff
@@ -177,6 +271,44 @@ def registered(self, required, provided, name=u''):
 
         return components.get(name)
 
+    @classmethod
+    def _allKeys(cls, components, i, parent_k=()):
+        if i == 0:
+            for k, v in components.items():
+                yield parent_k + (k,), v
+        else:
+            for k, v in components.items():
+                new_parent_k = parent_k + (k,)
+                for x, y in cls._allKeys(v, i - 1, new_parent_k):
+                    yield x, y
+
+    def _all_entries(self, byorder):
+        # Locally reference the `byorder` data; it might be replaced while
+        # this method is running.
+        for i, components in enumerate(byorder):
+            # We will have *i* levels of dictionaries to go before
+            # we get to the leaf.
+            for key, value in self._allKeys(components, i + 1):
+                assert len(key) == i + 2
+                required = key[:i]
+                provided = key[-2]
+                name = key[-1]
+                yield (required, provided, name, value)
+
+    def allRegistrations(self):
+        """
+        Yields tuples ``(required, provided, name, value)`` for all
+        the registrations that this object holds.
+
+        These tuples could be passed as the arguments to the
+        :meth:`register` method on another adapter registry to
+        duplicate the registrations this object holds.
+
+        .. versionadded:: 5.3.0
+        """
+        for t in self._all_entries(self._adapters):
```

Comment:
If this is Python 3 only it might be written as:
```
yield from self._all_entries(self._adapters)
```

Reply:
> If this is Python 3 only

It is not. The core libraries (zope.toolkit, perhaps, plus Acquisition?) are still supporting Python 2.7.